lib/multitask.py

```diff
@@ -0,0 +1,455 @@
+
+import tensorflow as tf
+import tensorflow.keras as tkf
+import lib.weighted_resblock as wblock
+import lib.shared_resnet as sresnet
+import lib.custom_resnet as cresnet
+from utils import args_util, training, datasets_util
+
+def get_custom_parser():
+  """returns a customized arguments parser."""
+  parser = args_util.get_parser()
+  parser.add_argument("--num_datasets",
+                      type=int,
+                      default=1,
+                      help="number of datasets in pretraining phase.")
+  parser.add_argument("--target_dataset",
+                      type=str,
+                      default="svhn_cropped",
+                      help="number of datasets in pretraining phase.")
+  parser.add_argument("--dataset_type",
+                      choices=["digits_da", "digits",
+                               "characters", "small_natural", "natural",
+                               "mix", "mix_2", "office", "domain_net"],
+                      default="digits",
+                      help="dataset type (default is digits)")
+  parser.add_argument("--share_mixture",
+                      type=int,
+                      default=0,
+                      help="if > 0, the mixture weights are shared.")
+  parser.add_argument("--share_logits",
+                      type=int,
+                      default=0,
+                      help="if > 0, the last layer is shared across datasets.")
+  parser.add_argument("--num_epochs_finetune",
+                      type=int,
+                      default=10,
+                      help="number of epochs to finetune the model.")
+  parser.add_argument("--finetune_mode",
+                      choices=["h", "m", "b", "hm", "hb", "hbm", "all"],
+                      default="h",
+                      help="finetuning mode "
+                      "(h = head, m = mixture weights,"
+                      "b = batch norm, all = entire model.)")
+  return parser
+
+def get_losses(datasets, num_train_ds, label_smoothing=0):
+  """Returens dictionaries of losses and loss weights for the datasets.
+
+  Arguments:
+  datasets: list of dataset names.
+  num_train_ds: number of datasets with non-zero loss weight.
+  label_smoothing: label smoothing coefficient.
+  """
+  cce = tf.keras.losses.CategoricalCrossentropy(label_smoothing=label_smoothing)
+  loss_weights = {}
+  losses = {}
+  for i in range(len(datasets)):
+    loss_name = "%s_out" % datasets[i]
+    loss_weights[loss_name] = float(i < num_train_ds)
+    losses[loss_name] = cce
+  return losses, loss_weights
+
+def get_dataset_list(dataset_type, num_datasets=None):
+  """Maps the dataset type with the list of dataset names.
+
+  Arguments:
+  dataset_type: datasets type.
+  num_datasets: optional number of datasets to use.
+  """
+  switcher = {
```

Comment:
I would move this outside of the function into the global scope (with a more specific name too).